stetho/src/main/java/com/facebook/stetho/screencast/android/ScreencastDispatcher.java

```diff
@@ -0,0 +1,151 @@
+      mHandlerThread = new HandlerThread("Screencast Thread");
+      mHandlerThread.start();
+    }
+    mPeer = peer;
+    mIsRunning = true;
+    mStream = new ByteArrayOutputStream();
+    mBackgroundHandler = new Handler(mHandlerThread.getLooper());
+    mMainHandler.postDelayed(mBitmapFetchRunnable, FRAME_DELAY);
+  }
+
+  public void stopScreencast() {
+    LogUtil.d("Stopping screencast");
+    mMainHandler.removeCallbacks(mBitmapFetchRunnable);
+    mBackgroundHandler.removeCallbacks(mEventDispatchRunnable);
+    mIsRunning = false;
+    mBitmap = null;
+    mCanvas = null;
+    mStream = null;
+  }
+
+  private class BitmapFetchRunnable implements Runnable {
+    @Override
+    public void run() {
+      getScreenBitmap();
+      mBackgroundHandler.post(mEventDispatchRunnable.withEndAction(this));
+    }
+
+    private Bitmap getScreenBitmap() {
+      if (!mIsRunning) {
+        return null;
+      }
+      Activity activity = mActivityTracker.tryGetTopActivity();
+      if (activity == null) {
+        return null;
+      }
+      // This stuff needs to happen in the UI thread
+      View rootView = activity.getWindow().getDecorView();
+      try {
+        if (mBitmap == null) {
+          int viewWidth = rootView.getWidth();
+          int viewHeight = rootView.getHeight();
+          float scale = Math.min((float) mRequest.maxWidth / (float) viewWidth,
+              (float) mRequest.maxHeight / (float) viewHeight);
+          int destWidth = (int) (viewWidth * scale);
+          int destHeight = (int) (viewHeight * scale);
+          mBitmap = Bitmap.createBitmap(destWidth, destHeight, Bitmap.Config.RGB_565);
+          mCanvas = new Canvas(mBitmap);
+          Matrix matrix = new Matrix();
+          mTempSrc.set(0, 0, viewWidth, viewHeight);
+          mTempDst.set(0, 0, destWidth, destHeight);
+          matrix.setRectToRect(mTempSrc, mTempDst, Matrix.ScaleToFit.CENTER);
+          mCanvas.setMatrix(matrix);
+        }
+        rootView.draw(mCanvas);
+        return mBitmap;
+      } catch (OutOfMemoryError e) {
+        LogUtil.w("Out of memory trying to allocate screencast Bitmap.");
+        return null;
+      }
+    }
+  }
+
+  private class EventDispatchRunnable implements Runnable {
+    private Runnable mEndAction;
+
+    private EventDispatchRunnable withEndAction(Runnable endAction) {
+      mEndAction = endAction;
+      return this;
+    }
+
+    @Override
+    public void run() {
+      if (!mIsRunning || mBitmap == null) {
+        return;
+      }
+      int width = mBitmap.getWidth();
+      int height = mBitmap.getHeight();
+      mStream.reset();
+      Base64OutputStream base64Stream = new Base64OutputStream(mStream, Base64.DEFAULT);
+      // request format is either "jpeg" or "png"
+      Bitmap.CompressFormat format = Bitmap.CompressFormat.valueOf(mRequest.format.toUpperCase());
+      mBitmap.compress(format, mRequest.quality, base64Stream);
+      mEvent.data = mStream.toString();
```

Comment:
I'm seeing a NullPointerException happening here, and it's easy to reproduce. There's a race between `stopScreencasting()` and this method: `mStream` is non-null when we call its `reset()` method, and then it's null by the time we call its `toString()` method. There are various ways to fix this. One possibility is that this method can load all the values that it needs from the outer class into local variables and then only use those. Another possibility is to implement a better cancellation system, e.g. `stopScreencasting()` would enqueue a `Runnable` to `mHandlerThread` whose job would be to tear everything down (e.g. set fields to null and stuff). That would guarantee that these don't have races.